apps/ejabberd/src/mod_aws_sns.erl

```diff
@@ -0,0 +1,293 @@
+                             Packet :: jlib:xmlel()) -> mod_aws_sns:attributes().
+
+%%%===================================================================
+%%% Types and definitions
+%%%===================================================================
+
+-define(TOPIC_BASE, ["arn", "aws", "sns"]).
+
+-type user_guid() :: binary().
+-type topic_arn() :: string(). %% Full topic ARN in format arn:aws:sns:{REGION}:{ACCOUNT_ID}:{TOPIC}
+-type topic() :: string(). %% {TOPIC} part of topic_arn() type
+-type attributes() :: #{string() => string() | binary() | number()}.
+
+
+%%%===================================================================
+%%% Exports
+%%%===================================================================
+
+%% MIM module callbacks
+-export([start/2, stop/1]).
+
+%% Hooks
+-export([user_send_packet/3, user_present/1, user_not_present/4, filter_room_packet/2]).
+
+%% API
+-export([try_publish/5]).
+
+%% Types
+-export_type([user_guid/0, topic_arn/0, topic/0, attributes/0]).
+
+%%%===================================================================
+%%% API functions
+%%%===================================================================
+
+-spec start(Host :: ejabberd:server(), Opts :: proplists:proplist()) -> ok.
+start(Host, Opts) ->
+    MUCHost = gen_mod:get_opt_subhost(Host, muc_host, Opts, mod_muc:default_host()),
+    ejabberd_hooks:add(filter_room_packet, MUCHost, ?MODULE, filter_room_packet, 90),
+    ejabberd_hooks:add(rest_user_send_packet, Host, ?MODULE, user_send_packet, 90),
+    ejabberd_hooks:add(user_send_packet, Host, ?MODULE, user_send_packet, 90),
+    ejabberd_hooks:add(user_available_hook, Host, ?MODULE, user_present, 90),
+    ejabberd_hooks:add(unset_presence_hook, Host, ?MODULE, user_not_present, 90),
+
+    application:ensure_all_started(erlcloud),
+    application:ensure_all_started(worker_pool),
+
+    WorkerNum = gen_mod:get_opt(pool_size, Opts, 100),
+    wpool:start_sup_pool(pool_name(Host), [{workers, WorkerNum}]),
+
+    %% Check for required options
+    RequiredOptions = [access_key_id, secret_access_key, region, account_id, sns_host],
+    lists:foreach(
+        fun(Option) ->
+            case opt(Host, Option) of
+                undefined ->
+                    error({missing_required_option, Option});
+                _ -> ok
+            end
+         end, RequiredOptions),
+
+    ok.
+
+-spec stop(Host :: ejabberd:server()) -> ok.
+stop(Host) ->
+    MUCHost = gen_mod:get_module_opt_subhost(Host, ?MODULE, mod_muc:default_host()),
+    ejabberd_hooks:delete(unset_presence_hook, Host, ?MODULE, user_not_present, 90),
+    ejabberd_hooks:delete(user_available_hook, Host, ?MODULE, user_present, 90),
+    ejabberd_hooks:delete(user_send_packet, Host, ?MODULE, user_send_packet, 90),
+    ejabberd_hooks:delete(rest_user_send_packet, Host, ?MODULE, user_send_packet, 90),
+    ejabberd_hooks:delete(filter_room_packet, MUCHost, ?MODULE, filter_room_packet, 90),
```

Comment:
Well, I did try, maybe because I spent most of the last month trying to untagle the terrible mess of hooks and handlers in the legacy code... But it shouldn't be a big deal - e.g. when `mod_muc_room` gets a go-aheas from `filter_room_packet` then, just before routing it out, it could call another hook, say `muc_send_packet`, and this'd be the one for you to use. Easy, cheap, and so much more readable.